docs/internal_abi.md

```diff
@@ -0,0 +1,323 @@
+In Verona; however, no mutable objects are ever shared and so, aside from reference count manipulations on immutable objects and interactions between adjacently allocated objects in different regions, there are no negative interactions with modern caches.
+The Verona field layout algorithm is therefore free to optimise for space.
+
+All fields whose size is a power-of-two should be sorted by size and laid out largest to smallest.
+The end of each section is more strongly aligned than the next requires and so this does not require any padding.
+
+Next, any union types that have a power-of-two sized payload but a smaller discriminator should be split into payload and discriminator.
+The payloads should be inserted into the sections for the same sized power-of-two objects.
+The discriminators should be packed into a single larger value.
+
+*Rationale:* Extracting from a bitfield is no more expensive if the bitfield is densely packed and if a function is pattern matching on multiple fields then they may be able to reuse a register for the descriptor values.
+
+The remaining values should be stored in the order in which they are declared, between the power-of-two region and the descriptors.
+
+*Rationale:* We could do some clever bin packing here but systems programmers who are using odd-sized storage-only types are likely to be careful about arranging them usefully already.
+A future version can improve bin packing and not break anything.
+
+
+*Note:* Java implementations typically pack all pointers together such that the trace logic just needs to look at a range, rather than have data or code describing the complete layout of the object.
+This may be possible for Verona, depending on the constraints for value types as fields and on pointers to value types.
+If value types cannot be passed by reference then they do not need to be stored contiguously in fields, because they will always be either copied when passed down the stack or accessed from functions that are aware of the location of their components.
+
+External references
+-------------------
+
+External references are structures with the same header as a Verona object followed by two pointers, one to the table that stores the external reference and one to the object.
+The rules for operations on this are defined in [`externalreference.h`](../src/rt/region/externalreference.h).
+External references are never embedded in other structures, they are always referenced by pointer and so fields or arrays of external references follow the same rules as any other kind of object pointer.
+
+Arrays
+------
+
+Arrays are objects (with the standard header) followed by a run of values of the same type.
+
+*Open question*: Should `Array[Bool]` be equivalent to `Array[U1]` or `Array[U8]` in storage?
+
+*Open question*: Should descriptors for array elements be collected in the same way as for structures?
```

Comment:
This means an unbound number of descriptors, which can only be packed at runtime. Also, if you add them to the beginning (or the end) of arrays, you end up having to read two cache pages for elements far from the table.

You could pack them at intervals, for example, as they accumulate into the size of the element itself, so you have a simple algorithm without breaking the alignment and caching locality, but making random access slightly harder. Though, if the interval is regular, it could be simple arithmetic.

Something like: `obj | obj | obj | obj | desc | obj | ...`, both `obj` and `desc` with the same alighment.

Reply:
I'm assuming something along the lines you describe.  It may not be something that actually matters, but something like `Array[U64 | T*]` is annoying because you effectively have an array of 65-bit values that require 64-bit alignment.  Your choices in representation are either 16 bytes per element or trying to collect the descriptor bits.  Having one 64-bit word of descriptors per 64 array elements isn't too bad from the perspective of iteration (you'll just grab it once and then if you're matching in the loop body then you'll shift and branch on the low one) but it sucks for random access because now you need to load two cache lines to get a single value.

It might be better to give programmers predictable bad behaviour than unpredictable slightly better (for some value of better) behaviour.